#164691221 Add endpoint for creating an article

richien opened the pull request:
#### What does this PR do?

Add the endpoint `/api/articles/` for creating an article.

#### Description of Task to be completed?

- Create the articles shell app and integrate it to authors
- Make and migrate the database schema for articles 
- Create the Article model, serializer, renderer and view
- Add test cases for the Article model, serializer, and view
- Fix a misconfiguration in django_nose that was excluding model
test cases from being run in the coverage tests

#### How should this be manually tested?

- After cloning this repository, switch to this branch and set up the application locally,
- After running the migrations using `python manage.py migrate` ,  start the development server
by running the command `python manage.py runserver`.

- Open a REST client like Postman and create a POST request to `http://localhost:8000/api/articles/`
with the following format as the body:
 ```
 {  "article": {
                "title": "Lorem ipsum dolor ",
                "description": "Lorem ipsum dolor sit amet, consectetur ",
                "body": "Lorem ipsum dolor sit amet, metus, dapibus at porttitor vel,
                               vulputate sed nibh.Cras malesuada dui mi, sit amet pretium 
                               lacus pretium et. "
		}
}
```
- Before posting the request, add an authorization header to your request by selection `Bearer Token` under the `Authorization` menu on Postman and paste your token in the `Token` field.
- A successful request will generate a response with the following format:
```
{  "article": {
        "title": "Lorem ipsum dolor",
        "description": "Lorem ipsum dolor sit amet, consectetur",
        "body": "Lorem ipsum dolor sit amet, metus, dapibus at porttitor vel,
                       vulputate sed nibh.Cras malesuada dui mi, sit amet pretium 
                       lacus pretium et.",
        "slug": "lorem-ipsum-dolor-4889c072b99f3bd4cea1",
        "createdAt": "2019-04-03 21:31:52.291229+00:00",
        "updatedAt": "2019-04-03 21:31:52.291281+00:00",
        "favorited": false,
        "favoritesCount": 0,
        "author": {
            "username": "test"
        }
    }
}
``` 

#### Any background context you want to provide?

- There was a misconfiguration in django_nose that was excluding model
test cases from being run in the coverage tests. This is the resource I used to
resolve this: https://github.com/django-nose/django-nose/issues/180
- The **response** is meant to include the following information about the
**author** of the article:
```
"author": {
      "username": "jake",
      "bio": "I work at statefarm",
      "image": "https://i.stack.imgur.com/xHWG8.jpg",
      "following": false
    }
```
**Note** that the `bio`, `image`, and `following` fields would be retrieved from the **Profile**, which is not yet implemented, hence the partial response.
- Also, **note** that because the `Tag` model is not yet implemented, the request and response from this endpoint omits the `tagList` field. Refer to the [README](https://github.com/aqua-organisation/ah-django-Aqua/blob/develop/README.md) file for more on the expected format for requests and responses.

#### What are the relevant Pivotal Tracker stories?

 [#164691221](https://www.pivotaltracker.com/story/show/164691221)

#### Screenshots (if appropriate)
- The **request**
![request](https://user-images.githubusercontent.com/6644707/55620999-fea52680-57a4-11e9-836d-55eac0395d8f.PNG)


- The **response**
![response](https://user-images.githubusercontent.com/6644707/55621007-0795f800-57a5-11e9-919a-694ccecbdd77.PNG)

richien committed:
feat(create article): add endpoint for creating an article

- create the articles shell app and integrate it to authors
- make and migrate the database schema for articles and tags
- create the Article model, serializer,renderer and view
- add testcases for the Article model, serializer, and view
- fix a misconfiguration in django_nose that was excluding model
testcases from being run in the coverage tests

[Finishes #164691221]

reifred commented:
![article](https://user-images.githubusercontent.com/22131999/55540022-b6193a80-56ca-11e9-80d9-b58d98af00cf.PNG)
Thanks, @richien for the good work done. Though when I try to test the feature as directed I get this error. I don't know where I am going wrong. Yet the tests are passing.

reifred reviewed (changes requested):
Thanks for the work @richien. Though I request some changes in your PR description before running the server. I guess one has to first sync the database with the new Models by running 
```python manage.py migrate```. This is because I was facing an error before doing so. Otherwise, everything was running as expected

RonKbS reviewed (changes requested):
Kindly redo the migrations such that you have one migration file instead of two.
You can;
- Switch to develop, run `python manage.py migrate`.
- Switch back to your branch and after removing these files, rerun the commands to make migrations and apply them.
- One file should be generated following this.


Also, there is a story to add tags so you should remove the implementation here, unless you have a valid reaseon for including it.

NicholusMuwonge reviewed (approved):
This works fine . Thanks for the efforts @richien

richien commented:
> 
> 
> Thanks for the work @richien. Though I request some changes in your PR description before running the server. I guess one has to first sync the database with the new Models by running
> `python manage.py migrate`. This is because I was facing an error before doing so. Otherwise, everything was running as expected

Thank you for the feedback. This has been implemented.

richien commented:
> 
> 
> Kindly redo the migrations such that you have one migration file instead of two.
> You can;
> 
>     * Switch to develop, run `python manage.py migrate`.
> 
>     * Switch back to your branch and after removing these files, rerun the commands to make migrations and apply them.
> 
>     * One file should be generated following this.
> 
> 
> Also, there is a story to add tags so you should remove the implementation here, unless you have a valid reaseon for including it.

Thank you for the feedback. In regards to the tags, the description of what a request from this endpoint should look like was taken from the README file as shown below: 
![create-article](https://user-images.githubusercontent.com/6644707/55605505-86773a80-577d-11e9-91f5-ef0945a03609.PNG)
This why I implemented the feature this way. You will also notice that a tag is only created if it does not already exist. The reason for this implemention is so that a tag can either be created at the same time the article is created or it can be created via the tags endpoint and added to an article. That was my understanding of the feature based on the information in the README file. kindly let me know if that is the right implementation.

RonKbS commented:
> This why I implemented the feature this way. You will also notice that a tag is only created if it does not already exist. The reason for this implemention is so that a tag can either be created at the same time the article is created or it can be created via the tags endpoint and added to an article. That was my understanding of the feature based on the information in the README file. kindly let me know if that is the right implementation.

I hear and understand you, @richien.

However, going back to the story on the PT board, tags are not within its scope and so I request that you remove the implementation for tags.
When doing the story for tags, the fields required will be added as needed.

richien commented:
> 
> 
> > This why I implemented the feature this way. You will also notice that a tag is only created if it does not already exist. The reason for this implemention is so that a tag can either be created at the same time the article is created or it can be created via the tags endpoint and added to an article. That was my understanding of the feature based on the information in the README file. kindly let me know if that is the right implementation.
> 
> I hear and understand you, @richien.
> 
> However, going back to the story on the PT board, tags are not within its scope and so I request that you remove the implementation for tags.
> When doing the story for tags, the fields required will be added as needed.

Ok, let me implement this

richien commented:
> 
> 
> > > This why I implemented the feature this way. You will also notice that a tag is only created if it does not already exist. The reason for this implemention is so that a tag can either be created at the same time the article is created or it can be created via the tags endpoint and added to an article. That was my understanding of the feature based on the information in the README file. kindly let me know if that is the right implementation.
> > 
> > 
> > I hear and understand you, @richien.
> > However, going back to the story on the PT board, tags are not within its scope and so I request that you remove the implementation for tags.
> > When doing the story for tags, the fields required will be added as needed.
> 
> Ok, let me implement this

This has been done. Kindly share feedback on this implementation.

RonKbS commented on authors/apps/articles/models.py:
```diff
@@ -0,0 +1,52 @@
+        unique_slug = slug
+        # Use a hash string to create a unigue extension to
+        # append to the slug inorder to create a unique slag
+        extention = secrets.token_hex(10)
```
This might be an overkill if we're looking at articles with similar slugs.
I would like for you to implement an increase in number instead.
If we have two articles with a similar slug, let the second article's slug have an increment of 1 instead.

This will make it easier when we're looking at the `GET` feature for articles and have to use slugs as identifiers.

RonKbS commented on authors/apps/articles/serializers.py:
```diff
@@ -0,0 +1,65 @@
+        body = data.get('body', None)
+        author_id = data.get('author_id', None)
+
+        if title is None:
```
The number of if statements in this section are two too many.
Please reimplement this in such a way that you have one `serializers.ValidationError` and the last word in `An article must have a ___` is picked up dynamically.

richien replied:
Thank you for the feedback. While refactoring, I realized that these validation checks are only enforced if the attributes of the serializer fields are set to `required=False`. I also realized that these checks are redundant since the serializer already does these validations by default. Therefore, in an attempt to remain DRY, I discarded these checks and only use the validate method for adding an author to an article.

RonKbS commented on authors/apps/articles/tests/test_article_model.py:
```diff
@@ -0,0 +1,49 @@
+class ArticleTestCase(TestCase):
+    """ This class defines the test suite for the article model """
+
+    def setUp(self):
```
There are tests that have similar creations of users in the authentication folder.
Please put all data in setup methods in one base class that is inherited by all current and newly created test classes.

RonKbS reviewed (changes requested):
Awesome work @richien.

Kindly implement the changes requested so I can merge this PR.

richien commented:
> 
> 
> Awesome work @richien.
> 
> Kindly implement the changes requested so I can merge this PR.

Thank you for the review. All the requested changes have been implemented.
